### PR TITLE
SCM improvements

### LINKED ISSUE #4732: [scm] making scm_folder.txt optimisation usable for export + install
I am currently modernizing my conan recipes and replace my old `scm` implementation with the upstream version. While playing around with the new recipe, I discovered that the `scm_folder.txt` optimization did not work for some of my build scripts. In particular, since #4218, the optimization only triggers for conan **create** and not when using conan **export** followed by conan **install**. (By the way, the tip at the end of the [documentation](https://docs.conan.io/en/1.6/creating_packages/package_repo.html#capturing-the-remote-and-commit-from-git-scm-experimental) should probably be updated to make it clear that only **create** is currently supported by this optimization.)

Using the separated **export** and **install** commands is useful in larger monorepo builds where I can first export all the conan sub packages before building (i.e., via conan **install** --build=outdated ...) the root package. Conan figures out the correct build order and builds all packages as needed.

While I absolutely agree that #4218 had to be fixed, I am now wondering if it is possible to also support the **export**+**install** flow, in addition to the **create** only flow, in a sane and correct way.

Current Logic
-------------
Currently the work flow seems to be that, as soon as either the `revision` or `url` is captured by `scm`, `scm_folder.txt` is written by both, the export functionality of the **export** and **create** commands. However, the information in `scm_folder.txt` is actually only used during the package installation step of the **create** command. 

Alternative Idea
----------------
Instead of writing the current path into the `scm_folder.txt` we could directly copy the content of the directory into the source folder of the package during the export step (i.e., triggered either as part of the **export** or **create** command). Subsequent **install** commands can directly use this source and do not have to clone the repository again. Similarly, in the workflow that only uses the **create** command, the source is already present and no additional download is required.

In summary, when using the **create** command, exactly the same amount of work (i.e., one copy into the source folder) would be performed. The only difference is that we perform this copy in the export instead of the install step of the command.

What do you think, would this be approach be feasible? @jgsogo (since you just recently worked with this code) did I miss important aspects?

### LINKED ISSUE #5126: [suggestion] `scm` feature, for dirty builds to use local source
To help us debug your issue please explain:

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

```
Conan version 1.14.3
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.5 LTS
Release:	16.04
Codename:	xenial
```

See for context https://github.com/conan-io/conan/issues/4600

This is while using using the scm feature.
When local changes are made (source is not pristine), "last" commit revision is captured.
This is a problem when trying to rebuild dependencies accurately from source.
Source folder would be removed, and conan would checkout the "last" captured commit from scm. 
This commit does not reflect the local changes.
This change allows the use of `revision: "None"`, which will attempt to build using the local source folder.

Proposed changes at https://github.com/Mark-Hatch-Bose/conan.
Not opening a pull request for the changes until this is signaled as a desired feature.

Example usage:
```
def mark_dirty():
  git = tools.Git()
  if git.is_pristine():
    return "auto"
  else:
    return "None"

class TestRepo(ConanFile):
    scm = {
        "type": "git",
        "url": "git@github.com:Org/TestRepo.git",
        "revision": mark_dirty()
    }
```

### LINKED ISSUE #5195: SCM auto mode improvements
This is a new issue to manage together #5126 and #4732 (probably some other).

**Current pains**

- `scm_folder.txt` is tricky and ugly (internal concern mostly).
- Since the usage of the local sources should be avoided when running a `conan install` command (see next section), could be still a good mechanism to take benefit of the optimization without taking the sources from the local folder but not cloning the repo. (@niosHD at https://github.com/conan-io/conan/issues/4732)
- A recipe containing `scm` with a real revision exported is dangerous if the folder was not pristine because the recipe could be uploaded and the CI could run something different from the expected. (@Mark-Hatch-Bose at https://github.com/conan-io/conan/issues/5126)

**Already solved pains to be kept**

- The local sources shouldn't be used in a `conan install --build XX` because it could take sources too much changed or belonging even to a different version. It should use the sources only for `conan create`. Solved at #4218

**Proposition**

- The `scm` auto will copy the sources from the local directory to the `source` cache directory. It makes `scm_folder.txt` not needed (@niosHD)
- The `scm` auto won't replace the revision if the repo is not pristine. `auto` will be kept.
- The `conan upload` command won't allow uploading a recipe with `revision=auto`.
- The `conan install` will work out of the box, because if the `source` folder exists, it will just use it, without trying to use changes from the local directory.
- The `conan export` and `conan create` will accept a `--scm-dirty` parameter to force capturing the revision (commit) even if the repo is not pristine. Why? if a CI build clone the repository, it is common that some file is added or modified because of the pipeline, but indeed you want to force that commit to be considered valid, so you can write your CI script with the `--scm-dirty` because you know what are you doing.

Feedback, please! Let's target `1.17` for this. @jgsogo feedback welcome.

### PR DESCRIPTION
**Pull request by lasote**
Changelog: Feature: The `scm` feature does not replace the `scm.revision="auto"` field with the commit when uncommitted changes unless ``--scm-dirty`` argument is specified. The recipe in the local cache will be kept with `revision=auto`.

Changelog: Feature: The `conan upload` command forbids to upload a recipe that uses the `scm` feature containing `revision=auto` or `url=auto`, unless `--force` is used. 

Changelog: Feature: The `scm` feature captures the local sources in the local cache during the export, avoiding later issues of modified local sources.

Docs: https://github.com/conan-io/docs/pull/XXXX

Closes #5195  
Closes #5126  
Closes #4732

#PYVERS: Macos@py27, Windows@py36, Linux@py27, py34
#TAGS: svn, slow
#REVISIONS: 1

### CODE REVIEW

**Comment by lasote** [comment]
This issue cannot be solved marking the source_folder with a file (to know that the source has to e executed), because in case of a failed source() method, because of the dirty flag, it will remove also the scm contents. I think we don't have another choice than a special folder for the scm captured sources.

**Comment by lasote** [comment]
Also, cannot be done using the `exports_sources` folder because it would be compressed and uploaded.

File: conans/client/cmd/export.py [hunk 1]
```diff
@@ -179,43 +190,51 @@ def cmd_export(app, conanfile_path, name, version, user, channel, keep_source,
...
+
+    if not scm.is_pristine() and not ignore_dirty:
+        output.warn("There are uncommitted changes, skipping the replacement of 'scm.url' and "
+                    "'scm.revision' auto fields. Use --ignore-dirty to force it. The 'conan "
+                    "upload' command will prevent uploading recipes with 'auto' values in these "
+                    "fields.")
+        local_src_path = scm.get_local_path_to_url(scm_data.url) \
+            if scm_data.url != "auto" else conanfile_dir
+        return scm_data, local_src_path
+    else:
```
Comment (memsharded) [inline conans/client/cmd/export.py]: This ``else`` can be removed, and the structure of this function will be ``return`` based, like in the above ``if not captured``

File: conans/client/source.py [hunk 2]
```diff
@@ -155,38 +149,62 @@ def _clean_source_folder(folder):
...
+        return
+    dest_dir = os.path.normpath(os.path.join(src_folder, scm_data.subfolder or ""))
+    # In user space, if revision="auto", then copy
+    if scm_data.capture_origin or scm_data.capture_revision:  # FIXME: or clause?
+        scm = SCM(scm_data, conanfile_folder, output)
+        scm_url = scm_data.url if scm_data.url != "auto" else \
+            scm.get_qualified_remote_url(remove_credentials=True)
+
+        src_path = scm.get_local_path_to_url(url=scm_url)
+        if src_path:
```
Comment (memsharded) [inline conans/client/source.py]: Maybe check src_path != dest_dir?

File: conans/test/functional/scm/scm_test.py [hunk 3]
```diff
@@ -1,4 +1,4 @@
-import json
+import os
```
Comment (jgsogo) [inline conans/test/functional/scm/scm_test.py]:
```suggestion
```

**Pull request merged**